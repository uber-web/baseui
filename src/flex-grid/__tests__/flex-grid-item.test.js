/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
// @flow

import * as React from 'react';
import {mount} from 'enzyme';

import FlexGridItem, {
  flexGridItemMediaQueryStyle,
  flexGridItemStyle,
  getResponsiveValue,
} from '../flex-grid-item.js';
import {LightTheme} from '../../themes/index.js';

describe('FlexGridItem', () => {
  test('flexGridItemMediaQueryStyle', () => {
    const testProps = [
      {
        flexGridColumnCount: 2,
        flexGridColumnGap: 0,
        flexGridRowGap: 0,
        flexGridItemIndex: 0,
        flexGridItemCount: 4,
        snapshotName: '2 columns no gaps',
      },
      {
        flexGridColumnCount: 2,
        flexGridColumnGap: '1px',
        flexGridRowGap: '2px',
        flexGridItemIndex: 0,
        flexGridItemCount: 4,
        snapshotName: '2 columns with gaps',
      },
      {
        flexGridColumnCount: 2,
        flexGridColumnGap: '1px',
        flexGridRowGap: '2px',
        flexGridItemIndex: 1,
        flexGridItemCount: 4,
        snapshotName: '2 columns with gaps end of row',
      },
      {
        flexGridColumnCount: 2,
        flexGridColumnGap: '1px',
        flexGridRowGap: '2px',
        flexGridItemIndex: 2,
        flexGridItemCount: 4,
        snapshotName: '2 columns with gaps end of column',
      },
      {
        flexGridColumnCount: 2,
        flexGridColumnGap: '1px',
        flexGridRowGap: '2px',
        flexGridItemIndex: 3,
        flexGridItemCount: 4,
        snapshotName: '2 columns with gaps end of row and column',
      },
      {
        flexGridColumnCount: 2,
        flexGridColumnGap: '1px',
        flexGridRowGap: '2px',
        flexGridItemIndex: 2,
        flexGridItemCount: 3,
        snapshotName: '2 columns with gaps row ends early',
      },
    ];
    testProps.forEach(
      ({
        flexGridColumnCount,
        flexGridColumnGap,
        flexGridRowGap,
        flexGridItemIndex,
        flexGridItemCount,
        snapshotName,
      }) => {
        expect(
          flexGridItemMediaQueryStyle({
            flexGridColumnCount,
            flexGridColumnGap,
            flexGridRowGap,
            flexGridItemIndex,
            flexGridItemCount,
            $theme: LightTheme,
          }),
        ).toMatchSnapshot(snapshotName);
      },
    );
  });

  test('getResponsiveValue', () => {
    const testCases = [
      {testCase: [null, 0], expected: null},
      {testCase: [null, 1], expected: null},
      {testCase: ['10px', 0], expected: '10px'},
      {testCase: ['10px', 1], expected: '10px'},
      {testCase: [['10px', '20px'], 0], expected: '10px'},
      {testCase: [['10px', '20px'], 1], expected: '20px'},
      {testCase: [['10px', '20px'], 2], expected: '20px'},
    ];
    testCases.forEach(({testCase, expected}) => {
      expect(getResponsiveValue(...testCase)).toEqual(expected);
    });
  });

  test('flexGridItemStyle', () => {
    const snapshotProps = [
      // $FlowFixMe - intentionally missing props
      {
        snapshotName: 'no props',
      },
      {
        $flexGridColumnCount: 2,
        $flexGridColumnGap: '1px',
        $flexGridRowGap: '2px',
        $flexGridItemIndex: 0,
        $flexGridItemCount: 6,
        snapshotName: 'non-responsive',
      },
      {
        $flexGridColumnCount: [2, 3],
        $flexGridColumnGap: '1px',
        $flexGridRowGap: '2px',
        $flexGridItemIndex: 0,
        $flexGridItemCount: 6,
        snapshotName: 'responsive columns',
      },
      {
        $flexGridColumnCount: [2, 3],
        $flexGridColumnGap: ['1px', '2px'],
        $flexGridRowGap: ['2px', '4px'],
        $flexGridItemIndex: 0,
        $flexGridItemCount: 6,
        snapshotName: 'responsive columns and gaps',
      },
    ];
    snapshotProps.forEach(
      ({
        $flexGridColumnCount,
        $flexGridColumnGap,
        $flexGridRowGap,
        $flexGridItemIndex,
        $flexGridItemCount,
        snapshotName,
      }) => {
        expect(
          flexGridItemStyle({
            $flexGridColumnCount,
            $flexGridColumnGap,
            $flexGridRowGap,
            $flexGridItemIndex,
            $flexGridItemCount,
            $theme: LightTheme,
          }),
        ).toMatchSnapshot(snapshotName);
      },
    );
  });

  it('renders', () => {
    const wrapper = mount(<FlexGridItem />);
    expect(wrapper).toMatchSnapshot('with default styles');

    wrapper.setProps({overrides: {Block: {style: {color: 'red'}}}});
    expect(wrapper).toMatchSnapshot('with overridden styles');
  });
});
