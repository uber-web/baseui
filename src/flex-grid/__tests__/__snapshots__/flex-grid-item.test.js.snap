// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexGridItem flexGridItemMediaQueryStyle: {"flexGridColumnCount":1,"flexGridColumnGap":"1px","flexGridRowGap":"2px"} 1`] = `
Object {
  ":nth-child(1n-0)": Object {
    "marginBottom": "2px",
    "marginRight": 0,
  },
  ":nth-child(1n-0):last-child": Object {
    "marginRight": "calc(0 * (1px + (100% - (1 - 1) * 1px) / 1))",
  },
  ":nth-child(1n-0):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  "width": "calc((100% - (1 - 1) * 1px) / 1)",
}
`;

exports[`FlexGridItem flexGridItemMediaQueryStyle: {"flexGridColumnCount":1,"flexGridColumnGap":0,"flexGridRowGap":0} 1`] = `
Object {
  ":nth-child(1n-0)": Object {
    "marginBottom": "0px",
    "marginRight": 0,
  },
  ":nth-child(1n-0):last-child": Object {
    "marginRight": "calc(0 * (0px + (100% - (1 - 1) * 0px) / 1))",
  },
  ":nth-child(1n-0):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  "width": "calc((100% - (1 - 1) * 0px) / 1)",
}
`;

exports[`FlexGridItem flexGridItemMediaQueryStyle: {"flexGridColumnCount":2,"flexGridColumnGap":"1px","flexGridRowGap":"2px"} 1`] = `
Object {
  ":nth-child(2n-0)": Object {
    "marginBottom": "2px",
    "marginRight": 0,
  },
  ":nth-child(2n-0):last-child": Object {
    "marginRight": "calc(0 * (1px + (100% - (2 - 1) * 1px) / 2))",
  },
  ":nth-child(2n-0):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  ":nth-child(2n-1)": Object {
    "marginBottom": "2px",
    "marginRight": "1px",
  },
  ":nth-child(2n-1):last-child": Object {
    "marginRight": "calc(1 * (1px + (100% - (2 - 1) * 1px) / 2))",
  },
  ":nth-child(2n-1):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  ":nth-child(2n-1):nth-last-child(2)": Object {
    "marginBottom": 0,
  },
  "width": "calc((100% - (2 - 1) * 1px) / 2)",
}
`;

exports[`FlexGridItem flexGridItemMediaQueryStyle: {"flexGridColumnCount":3,"flexGridColumnGap":"1px","flexGridRowGap":"2px"} 1`] = `
Object {
  ":nth-child(3n-0)": Object {
    "marginBottom": "2px",
    "marginRight": 0,
  },
  ":nth-child(3n-0):last-child": Object {
    "marginRight": "calc(0 * (1px + (100% - (3 - 1) * 1px) / 3))",
  },
  ":nth-child(3n-0):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  ":nth-child(3n-1)": Object {
    "marginBottom": "2px",
    "marginRight": "1px",
  },
  ":nth-child(3n-1):last-child": Object {
    "marginRight": "calc(1 * (1px + (100% - (3 - 1) * 1px) / 3))",
  },
  ":nth-child(3n-1):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  ":nth-child(3n-1):nth-last-child(2)": Object {
    "marginBottom": 0,
  },
  ":nth-child(3n-2)": Object {
    "marginBottom": "2px",
    "marginRight": "1px",
  },
  ":nth-child(3n-2):last-child": Object {
    "marginRight": "calc(2 * (1px + (100% - (3 - 1) * 1px) / 3))",
  },
  ":nth-child(3n-2):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  ":nth-child(3n-2):nth-last-child(2)": Object {
    "marginBottom": 0,
  },
  ":nth-child(3n-2):nth-last-child(3)": Object {
    "marginBottom": 0,
  },
  "width": "calc((100% - (3 - 1) * 1px) / 3)",
}
`;

exports[`FlexGridItem flexGridItemStyle: {"$flexGridColumnCount":[1,2],"$flexGridColumnGap":"1px","$flexGridRowGap":"2px"} 1`] = `
Object {
  "@media screen and (max-width: 319px)": Object {
    ":nth-child(1n-0)": Object {
      "marginBottom": "2px",
      "marginRight": 0,
    },
    ":nth-child(1n-0):last-child": Object {
      "marginRight": "calc(0 * (1px + (100% - (1 - 1) * 1px) / 1))",
    },
    ":nth-child(1n-0):nth-last-child(1)": Object {
      "marginBottom": 0,
    },
    "width": "calc((100% - (1 - 1) * 1px) / 1)",
  },
  "@media screen and (min-width: 320px)": Object {
    ":nth-child(2n-0)": Object {
      "marginBottom": "2px",
      "marginRight": 0,
    },
    ":nth-child(2n-0):last-child": Object {
      "marginRight": "calc(0 * (1px + (100% - (2 - 1) * 1px) / 2))",
    },
    ":nth-child(2n-0):nth-last-child(1)": Object {
      "marginBottom": 0,
    },
    ":nth-child(2n-1)": Object {
      "marginBottom": "2px",
      "marginRight": "1px",
    },
    ":nth-child(2n-1):last-child": Object {
      "marginRight": "calc(1 * (1px + (100% - (2 - 1) * 1px) / 2))",
    },
    ":nth-child(2n-1):nth-last-child(1)": Object {
      "marginBottom": 0,
    },
    ":nth-child(2n-1):nth-last-child(2)": Object {
      "marginBottom": 0,
    },
    "width": "calc((100% - (2 - 1) * 1px) / 2)",
  },
  "flexGrow": 1,
}
`;

exports[`FlexGridItem flexGridItemStyle: {"$flexGridColumnCount":[1,2],"$flexGridColumnGap":["1px","2px"],"$flexGridRowGap":["2px","4px"]} 1`] = `
Object {
  "@media screen and (max-width: 319px)": Object {
    ":nth-child(1n-0)": Object {
      "marginBottom": "2px",
      "marginRight": 0,
    },
    ":nth-child(1n-0):last-child": Object {
      "marginRight": "calc(0 * (1px + (100% - (1 - 1) * 1px) / 1))",
    },
    ":nth-child(1n-0):nth-last-child(1)": Object {
      "marginBottom": 0,
    },
    "width": "calc((100% - (1 - 1) * 1px) / 1)",
  },
  "@media screen and (min-width: 320px)": Object {
    ":nth-child(2n-0)": Object {
      "marginBottom": "4px",
      "marginRight": 0,
    },
    ":nth-child(2n-0):last-child": Object {
      "marginRight": "calc(0 * (2px + (100% - (2 - 1) * 2px) / 2))",
    },
    ":nth-child(2n-0):nth-last-child(1)": Object {
      "marginBottom": 0,
    },
    ":nth-child(2n-1)": Object {
      "marginBottom": "4px",
      "marginRight": "2px",
    },
    ":nth-child(2n-1):last-child": Object {
      "marginRight": "calc(1 * (2px + (100% - (2 - 1) * 2px) / 2))",
    },
    ":nth-child(2n-1):nth-last-child(1)": Object {
      "marginBottom": 0,
    },
    ":nth-child(2n-1):nth-last-child(2)": Object {
      "marginBottom": 0,
    },
    "width": "calc((100% - (2 - 1) * 2px) / 2)",
  },
  "flexGrow": 1,
}
`;

exports[`FlexGridItem flexGridItemStyle: {"$flexGridColumnCount":2,"$flexGridColumnGap":"1px","$flexGridRowGap":"2px"} 1`] = `
Object {
  ":nth-child(2n-0)": Object {
    "marginBottom": "2px",
    "marginRight": 0,
  },
  ":nth-child(2n-0):last-child": Object {
    "marginRight": "calc(0 * (1px + (100% - (2 - 1) * 1px) / 2))",
  },
  ":nth-child(2n-0):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  ":nth-child(2n-1)": Object {
    "marginBottom": "2px",
    "marginRight": "1px",
  },
  ":nth-child(2n-1):last-child": Object {
    "marginRight": "calc(1 * (1px + (100% - (2 - 1) * 1px) / 2))",
  },
  ":nth-child(2n-1):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  ":nth-child(2n-1):nth-last-child(2)": Object {
    "marginBottom": 0,
  },
  "flexGrow": 1,
  "width": "calc((100% - (2 - 1) * 1px) / 2)",
}
`;

exports[`FlexGridItem flexGridItemStyle: {} 1`] = `
Object {
  ":nth-child(1n-0)": Object {
    "marginBottom": "0px",
    "marginRight": 0,
  },
  ":nth-child(1n-0):last-child": Object {
    "marginRight": "calc(0 * (0px + (100% - (1 - 1) * 0px) / 1))",
  },
  ":nth-child(1n-0):nth-last-child(1)": Object {
    "marginBottom": 0,
  },
  "flexGrow": 1,
  "width": "calc((100% - (1 - 1) * 0px) / 1)",
}
`;

exports[`FlexGridItem renders: with default styles 1`] = `
<FlexGridItem>
  <Block
    as="div"
    data-baseweb="flex-grid-item"
    overrides={
      Object {
        "Block": Object {
          "style": [Function],
        },
      }
    }
  >
    <MockStyledComponent
      $as="div"
      $style={[Function]}
      data-baseweb="flex-grid-item"
    >
      <div
        data-baseweb="flex-grid-item"
        styled-component="true"
      />
    </MockStyledComponent>
  </Block>
</FlexGridItem>
`;

exports[`FlexGridItem renders: with overridden styles 1`] = `
<FlexGridItem
  overrides={
    Object {
      "Block": Object {
        "style": Object {
          "color": "red",
        },
      },
    }
  }
>
  <Block
    as="div"
    data-baseweb="flex-grid-item"
    overrides={
      Object {
        "Block": Object {
          "style": [Function],
        },
      }
    }
  >
    <MockStyledComponent
      $as="div"
      $style={[Function]}
      data-baseweb="flex-grid-item"
    >
      <div
        data-baseweb="flex-grid-item"
        styled-component="true"
      />
    </MockStyledComponent>
  </Block>
</FlexGridItem>
`;
